### PR TITLE
[12.0] ADD stock_picking_type_user_restriction

### PR DESCRIPTION
**Pull request by eLBati**
Restrict some users to only access their assigned operation types.

Same structure as [pos_user_restriction](https://github.com/OCA/pos/pull/514)

### CODE REVIEW

**Review by rousseldenis — APPROVED** [review]
Code review. LGTM.

In a Roadmap, you could depends on https://github.com/OCA/server-tools/tree/10.0/base_user_role

And add a link to 'role_ids' in picking types as it can be tough to add or remove people if you have a lot of picking types and users.

**Review by HaraldPanten — CHANGES_REQUESTED** [review]
I think that maybe there's something missing:

1- I set demo user to "User: Assigned Operation Types Only" group.

2- I log in with demo user.

**issue:** I do not have access to inventory module. Shouldn't I have permissions enough with this user to access inventory module?

IMO, you shouldn't restrict inventory module access.

**Comment by eLBati** [comment]
@HaraldPanten right.
Added the same permissions as stock user, restricting to assigned operation types.
Thanks

**Review by mrcast — COMMENTED** [review]
@eLBati 
as Admin 
1) I set demo user to "User: Assigned Operation Types Only" group.
2) I set demo user on a operation type (receipts - yourcompany)
3) I log with demo user and  I can see in dashboard all the operation types with no restrictions

I'm forgetting something?

**Comment by eLBati** [comment]
@mrcast if you set demo user on an operation type, demo user will be able to see that operation type. Try setting another user

**Review by mrcast — APPROVED** [review]
@eLBati : funtional test ok!

**Review by HaraldPanten — APPROVED** [review]
LGTM 👍

**Review by marcelsavegnago — APPROVED** [review]

File: stock_picking_type_user_restriction/tests/test_picking_type_user_restriction.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        self.stock_user_assigned_type = self.env['res.users'].create({
+            'login': 'stock_user_assigned_type',
+            'name': 'stock_user_assigned_type',
+            'groups_id': [(6, 0, [self.env.ref(
+                'stock_picking_type_user_restriction.'
+                'group_assigned_picking_types_user'
+            ).id])]
+        })
+        self.picking_type_out = self.env.ref('stock.picking_type_out')
+        self.picking_type_model = self.env['stock.picking.type']
```
Comment (simahawk) [inline stock_picking_type_user_restriction/tests/test_picking_type_user_restriction.py]: @eLBati pls never create users in tests w/out `no_reset_password` (it generates emails for pwd!) and when possible use `tracking_disable` and SavepointCase ;)

```suggestion
from odoo.tests.common import SavepointCase


class TestUserRestriction(SavepointCase):

    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        cls.env = cls.env(context=dict(
            cls.env.context,
            tracking_disable=True,
            no_reset_password=True,
        ))
        cls.stock_user = cls.env['res.users'].create({
            'login': 'stock_user',
            'name': 'stock_user',
            'groups_id': [(6, 0, [cls.env.ref('stock.group_stock_user').id])]
        })
        cls.stock_user_assigned_type = cls.env['res.users'].create({
            'login': 'stock_user_assigned_type',
            'name': 'stock_user_assigned_type',
            'groups_id': [(6, 0, [cls.env.ref(
                'stock_picking_type_user_restriction.'
                'group_assigned_picking_types_user'
            ).id])]
        })
        cls.picking_type_out = cls.env.ref('stock.picking_type_out')
        cls.picking_type_model = cls.env['stock.picking.type']
```

**Review by simahawk — CHANGES_REQUESTED** [review]

File: stock_picking_type_user_restriction/tests/test_picking_type_user_restriction.py (hunk 1 above)
Comment (eLBati) [reply stock_picking_type_user_restriction/tests/test_picking_type_user_restriction.py]: Thanks, committed

**Comment by simahawk** [comment]
@eLBati nice commit message :smile: 
Build is red, can you check?

**Comment by eLBati** [comment]
> nice commit message

Thanks, my credits :nerd_face: 

> Build is red, can you check?

Relaunched and green

**Pull request merged**